File: android/app/src/main/java/com/justspent/app/utils/DateFilter.kt
package com.justspent.app.utils

import java.time.LocalDate
import java.time.LocalDateTime
import java.time.LocalTime
import java.time.format.DateTimeFormatter
import java.time.temporal.TemporalAdjusters
import kotlinx.datetime.LocalDateTime as KotlinxLocalDateTime

/**
 * Represents the available date filter options for expense lists
 */
sealed class DateFilter {
    /** No filter - show all expenses */
    data object All : DateFilter()

    /** Show only today's expenses */
    data object Today : DateFilter()

    /** Show expenses from the last 7 days (rolling) */
    data object Week : DateFilter()

    /** Show expenses from the current calendar month */
    data object Month : DateFilter()

    /** Show expenses within a custom date range */
    data class Custom(val start: LocalDate, val end: LocalDate) : DateFilter()

    /**
     * Display name for the filter
     */
    val displayName: String
        get() = when (this) {
            is All -> "All"
            is Today -> "Today"
            is Week -> "Week"
            is Month -> "Month"
            is Custom -> "Custom"
        }

    /**
     * Returns the formatted date range string for custom filter
     */
    val customRangeDisplayString: String?
        get() = when (this) {
            is Custom -> {
                val formatter = DateTimeFormatter.ofPattern("MMM d")
                "${start.format(formatter)} - ${end.format(formatter)}"
            }
            else -> null
        }
}

/**
 * Utility class for date filter calculations
 */
object DateFilterUtils {

    /**
     * Returns the date range for a given filter
     *
     * @param filter The date filter to get range for
     * @param referenceDate The reference date (defaults to today)
     * @return A pair of (startDateTime, endDateTime) or null for All filter
     */
    fun dateRange(
        filter: DateFilter,
        referenceDate: LocalDate = LocalDate.now()
    ): Pair<LocalDateTime, LocalDateTime>? {
        return when (filter) {
            is DateFilter.All -> null

            is DateFilter.Today -> {
                val startOfDay = referenceDate.atStartOfDay()
                val endOfDay = referenceDate.atTime(LocalTime.MAX)
                startOfDay to endOfDay
            }

            is DateFilter.Week -> {
                // Last 7 days (rolling)
                val endOfDay = referenceDate.atTime(LocalTime.MAX)
                val startOfWeek = referenceDate.minusDays(6).atStartOfDay()
                startOfWeek to endOfDay
            }

            is DateFilter.Month -> {
                // Current calendar month
                val startOfMonth = referenceDate.with(TemporalAdjusters.firstDayOfMonth()).atStartOfDay()
                val endOfMonth = referenceDate.with(TemporalAdjusters.lastDayOfMonth()).atTime(LocalTime.MAX)
                startOfMonth to endOfMonth
            }

            is DateFilter.Custom -> {
                val startOfDay = filter.start.atStartOfDay()
                val endOfDay = filter.end.atTime(LocalTime.MAX)
                startOfDay to endOfDay
            }
        }
    }

    /**
     * Validates a custom date range
     *
     * @param startDate The start date
     * @param endDate The end date
     * @param referenceDate The reference date for "future" check (defaults to today)
     * @return A validation result with any error messages
     */
    fun validateCustomRange(
        startDate: LocalDate,
        endDate: LocalDate,
        referenceDate: LocalDate = LocalDate.now()
    ): DateRangeValidationResult {
        val errors = mutableListOf<String>()

        // Check if end date is before start date
        if (endDate.isBefore(startDate)) {
            errors.add("End date must be on or after start date")
        }

        // Check if dates are in the future
        val tomorrow = referenceDate.plusDays(1)
        if (startDate.isAfter(referenceDate) || startDate.isEqual(tomorrow)) {
            errors.add("Start date cannot be in the future")
        }
        if (endDate.isAfter(referenceDate) || endDate.isEqual(tomorrow)) {
            errors.add("End date cannot be in the future")
        }

        // Check if range is too large (more than 1 year)
        val daysDifference = java.time.temporal.ChronoUnit.DAYS.between(startDate, endDate)
        if (daysDifference > 365) {
            errors.add("Date range cannot exceed 1 year")
        }

        return DateRangeValidationResult(
            isValid = errors.isEmpty(),
            errors = errors
        )
    }

    /**
     * Checks if a date falls within the filter's range
     *
     * @param dateTime The date/time to check (java.time.LocalDateTime)
     * @param filter The filter to check against
     * @param referenceDate The reference date (defaults to today)
     * @return True if the date is within the filter's range
     */
    fun isDateInFilter(
        dateTime: LocalDateTime,
        filter: DateFilter,
        referenceDate: LocalDate = LocalDate.now()
    ): Boolean {
        val range = dateRange(filter, referenceDate) ?: return true // All filter includes all dates

        return !dateTime.isBefore(range.first) && !dateTime.isAfter(range.second)
    }

    /**
     * Checks if a date falls within the filter's range (kotlinx.datetime version)
     *
     * @param dateTime The date/time to check (kotlinx.datetime.LocalDateTime)
     * @param filter The filter to check against
     * @param referenceDate The reference date (defaults to today)
     * @return True if the date is within the filter's range
     */
    fun isDateInFilter(
        dateTime: KotlinxLocalDateTime,
        filter: DateFilter,
        referenceDate: LocalDate = LocalDate.now()
    ): Boolean {
        // Convert kotlinx.datetime.LocalDateTime to java.time.LocalDateTime
        val javaDateTime = LocalDateTime.of(
            dateTime.year,
            dateTime.monthNumber,
            dateTime.dayOfMonth,
            dateTime.hour,
            dateTime.minute,
            dateTime.second,
            dateTime.nanosecond
        )
        return isDateInFilter(javaDateTime, filter, referenceDate)
    }
}

/**
 * Result of date range validation
 */
data class DateRangeValidationResult(
    /** Whether the range is valid */
    val isValid: Boolean,
    /** List of validation errors (empty if valid) */
    val errors: List<String>
) {
    /** Convenience property for first error */
    val firstError: String?
        get() = errors.firstOrNull()
}

File: ios/JustSpent/JustSpent/Models/DateFilter.swift
//
//  DateFilter.swift
//  JustSpent
//
//  Created by Claude Code on 2025.
//

import Foundation

/// Represents the available date filter options for expense lists
enum DateFilter: Equatable, Hashable {
    /// No filter - show all expenses
    case all
    /// Show only today's expenses
    case today
    /// Show expenses from the last 7 days (rolling)
    case week
    /// Show expenses from the current calendar month
    case month
    /// Show expenses within a custom date range
    case custom(start: Date, end: Date)

    /// Display name for the filter
    var displayName: String {
        switch self {
        case .all:
            return "All"
        case .today:
            return "Today"
        case .week:
            return "Week"
        case .month:
            return "Month"
        case .custom:
            return "Custom"
        }
    }

    /// Accessibility label for VoiceOver
    var accessibilityLabel: String {
        switch self {
        case .all:
            return "Show all expenses"
        case .today:
            return "Show today's expenses"
        case .week:
            return "Show this week's expenses"
        case .month:
            return "Show this month's expenses"
        case .custom(let start, let end):
            let formatter = DateFormatter()
            formatter.dateStyle = .medium
            return "Custom range from \(formatter.string(from: start)) to \(formatter.string(from: end))"
        }
    }

    /// Returns the formatted date range string for custom filter
    var customRangeDisplayString: String? {
        guard case .custom(let start, let end) = self else { return nil }
        let formatter = DateFormatter()
        formatter.dateFormat = "MMM d"
        return "\(formatter.string(from: start)) - \(formatter.string(from: end))"
    }

    // MARK: - Equatable

    static func == (lhs: DateFilter, rhs: DateFilter) -> Bool {
        switch (lhs, rhs) {
        case (.all, .all), (.today, .today), (.week, .week), (.month, .month):
            return true
        case (.custom(let lStart, let lEnd), .custom(let rStart, let rEnd)):
            return Calendar.current.isDate(lStart, inSameDayAs: rStart) &&
                   Calendar.current.isDate(lEnd, inSameDayAs: rEnd)
        default:
            return false
        }
    }

    // MARK: - Hashable

    func hash(into hasher: inout Hasher) {
        switch self {
        case .all:
            hasher.combine(0)
        case .today:
            hasher.combine(1)
        case .week:
            hasher.combine(2)
        case .month:
            hasher.combine(3)
        case .custom(let start, let end):
            hasher.combine(4)
            hasher.combine(start)
            hasher.combine(end)
        }
    }
}

/// Utility class for date filter calculations
class DateFilterUtils {

    /// Shared instance
    static let shared = DateFilterUtils()

    /// Calendar for date calculations
    private let calendar: Calendar

    init(calendar: Calendar = .current) {
        self.calendar = calendar
    }

    /// Returns the date range for a given filter
    /// - Parameters:
    ///   - filter: The date filter to get range for
    ///   - referenceDate: The reference date (defaults to now)
    /// - Returns: A tuple of (startDate, endDate) or nil for .all filter
    func dateRange(for filter: DateFilter, referenceDate: Date = Date()) -> (start: Date, end: Date)? {
        switch filter {
        case .all:
            return nil

        case .today:
            let startOfDay = calendar.startOfDay(for: referenceDate)
            let endOfDay = calendar.date(byAdding: .day, value: 1, to: startOfDay)!.addingTimeInterval(-1)
            return (startOfDay, endOfDay)

        case .week:
            // Last 7 days (rolling)
            let endOfDay = calendar.date(byAdding: .day, value: 1, to: calendar.startOfDay(for: referenceDate))!.addingTimeInterval(-1)
            let startOfWeek = calendar.date(byAdding: .day, value: -6, to: calendar.startOfDay(for: referenceDate))!
            return (startOfWeek, endOfDay)

        case .month:
            // Current calendar month
            let components = calendar.dateComponents([.year, .month], from: referenceDate)
            let startOfMonth = calendar.date(from: components)!
            var endComponents = DateComponents()
            endComponents.month = 1
            endComponents.second = -1
            let endOfMonth = calendar.date(byAdding: endComponents, to: startOfMonth)!
            return (startOfMonth, endOfMonth)

        case .custom(let start, let end):
            let startOfDay = calendar.startOfDay(for: start)
            let endOfDay = calendar.date(byAdding: .day, value: 1, to: calendar.startOfDay(for: end))!.addingTimeInterval(-1)
            return (startOfDay, endOfDay)
        }
    }

    /// Validates a custom date range
    /// - Parameters:
    ///   - startDate: The start date
    ///   - endDate: The end date
    ///   - referenceDate: The reference date for "future" check (defaults to now)
    /// - Returns: A validation result with any error messages
    func validateCustomRange(startDate: Date, endDate: Date, referenceDate: Date = Date()) -> DateRangeValidationResult {
        var errors: [String] = []

        // Check if end date is before start date
        if endDate < startDate {
            errors.append("End date must be on or after start date")
        }

        // Check if dates are in the future
        let tomorrow = calendar.date(byAdding: .day, value: 1, to: calendar.startOfDay(for: referenceDate))!
        if calendar.startOfDay(for: startDate) >= tomorrow {
            errors.append("Start date cannot be in the future")
        }
        if calendar.startOfDay(for: endDate) >= tomorrow {
            errors.append("End date cannot be in the future")
        }

        // Check if range is too large (more than 1 year)
        let daysDifference = calendar.dateComponents([.day], from: startDate, to: endDate).day ?? 0
        if daysDifference > 365 {
            errors.append("Date range cannot exceed 1 year")
        }

        return DateRangeValidationResult(isValid: errors.isEmpty, errors: errors)
    }

    /// Checks if a date falls within the filter's range
    /// - Parameters:
    ///   - date: The date to check
    ///   - filter: The filter to check against
    ///   - referenceDate: The reference date (defaults to now)
    /// - Returns: True if the date is within the filter's range
    func isDate(_ date: Date, inFilter filter: DateFilter, referenceDate: Date = Date()) -> Bool {
        guard let range = dateRange(for: filter, referenceDate: referenceDate) else {
            return true // .all filter includes all dates
        }
        return date >= range.start && date <= range.end
    }

    /// Creates an NSPredicate for filtering expenses by date
    /// - Parameters:
    ///   - filter: The date filter
    ///   - dateKeyPath: The key path for the date field (default: "transactionDate")
    ///   - referenceDate: The reference date (defaults to now)
    /// - Returns: An NSPredicate or nil for .all filter
    func predicate(for filter: DateFilter, dateKeyPath: String = "transactionDate", referenceDate: Date = Date()) -> NSPredicate? {
        guard let range = dateRange(for: filter, referenceDate: referenceDate) else {
            return nil
        }
        return NSPredicate(format: "%K >= %@ AND %K <= %@",
                          dateKeyPath, range.start as NSDate,
                          dateKeyPath, range.end as NSDate)
    }
}

/// Result of date range validation
struct DateRangeValidationResult {
    /// Whether the range is valid
    let isValid: Bool
    /// List of validation errors (empty if valid)
    let errors: [String]

    /// Convenience property for first error
    var firstError: String? {
        errors.first
    }
}
